### PR TITLE
Add a global switch to disable third party

### PR DESCRIPTION
**Pull request by eliecharra**
## Description

This PR is adding a more convenient way to disable any third-party network calls.

You can now build an telemetry-free version of driftctl easily with 
```shell
$ go build -ldflags="-X 'github.com/cloudskiff/driftctl/build.enableThirdPartyNetworkCalls=false'" 
```

## References
- https://github.com/NixOS/nixpkgs/commit/ded247ada384ceacc5d0d186cffa7c1cc927d268#diff-f6248b889160cf7652be9255c593b1ded04177a6d33449c69b8f4e4abf28ce57R31-R38

### CODE REVIEW

File: build/build.go
Comment (eliecharra) [inline build/build.go:9]: Please suggest me another naming, I'm not really convinced by that 😅

Comment (craigfurman) [reply build/build.go:9]: The awkward name might be suggesting that we've combined separate concepts into this one flag, as the comment explains. Use-case-wise, are disabling version checks and telemetry really the same thing?

Not that it's really a strong opinion of mine to separate this into 2 build flags, perhaps it's not worth it from a complexity PoV.

If we do want to keep them the same, how about: `enableTelemetry`, to lead with the thing users are more likely to care about disabling.

Comment (eliecharra) [reply build/build.go:9]: > The awkward name might be suggesting that we've combined separate concepts into this one flag

exactly 😅 

> Use-case-wise, are disabling version checks and telemetry really the same thing?

Some people reached us about that a couple of month ago, even if telemetry and version check can be opted-out, they wanted to bring a binary in their ecosystem free of that kind of code, so they forked and removed every external calls. 

Today it seems that nix pkg maintainers want to opt-out theses things by default.

So yeah to try to answer your question, I do not think that telemetry and version check are the same think, but both are - by default - leaking data in the internet, and here I think this is our usecase : be able to build a driftctl version free of any data leak.

Since dctl can be seen as a security tool, this should be taken seriously and I think "one flag to rule them all" is a better approach since this ensure maintainers that the version will be free of data leak in the time.

For example if we decide for some reason one day to enable by default crash reporting, we should be able to disabled it from our perspective without the need for the maintainers to check if there is a new data leak feat in every new version. 

> If we do want to keep them the same, how about: enableTelemetry, to lead with the thing users are more likely to care about disabling.

Maybe it's a good idea to consider everything that leak data as "telemetry" related 👍🏻. I am biased since we separated theses two concept in dctl, but maybe for most of the users it should be clear enough ?

Comment (moadibfr) [reply build/build.go:9]: Maybe something like enableUsageReport ? It's not 100% accurate but still it would fit with what we actually do with telemetry and version check and be clear enough to people that would use that ?
Third party network call sounds really strange to me.

Comment (craigfurman) [reply build/build.go:9]: @eliecharra that all sounds reasonable. Some possible alternative names: `enableTelemetryAndCheckup`, `enableTelemetryAndDiagnostics`? I'm being slightly more generic over the "version check" aspect

Comment (eliecharra) [reply build/build.go:9]: > Third party network call sounds really strange to me.

100% agree, whatever we choose this should not be this one 😅 

WDYT of `privateMode` ? It make me think of web browser private mode, I get the idea during the sleep.

Comment (sundowndev) [reply build/build.go:9]: May be we should simply use two different build arguments for this ? I've looked at how it's done in other projects and VSCode uses `enableCrashReporter` and `enableTelemetry` to disable those 3rd party calls. Also the user will better understand what they're disabling exactly, instead of being confused by a generic name. WDYT?

File: pkg/telemetry/telemetry.go
Comment (wbeuil) [inline pkg/telemetry/telemetry.go:36]: `Third party network callS are ...`

**Review by wbeuil — CHANGES_REQUESTED** [review]

File: build/build.go
Comment (eliecharra) [reply build/build.go:9]: I already answered to that question. BTW I do not think VSCode (or more generally Microsoft related products) are well designed about this, I would like to avoid these kind of things : https://vscodium.com/

Comment (wbeuil) [reply build/build.go:9]: I don't have a clear opinion on whether we should have 2 flags or 1 flag to disable/enable third-party calls. Both what Craig/Raph said to have different build args and what Elie said on "one flag to rule them all" kind of make sense to me.

Comment (wbeuil) [reply build/build.go:9]: But if we go the one flag road, I thought of `enableReporters` or both suggested names of Craig that are more transparent of what we disable/enable.

Comment (eliecharra) [reply build/build.go]: I switched it to `privateMode`, I think it fit well with what we want to achieve here.

Comment (craigfurman) [reply build/build.go]: Sorry to keep this thread going, and I'll defer to long-term maintainers on this, but: `privateMode` is perhaps a bit of a loaded term? It makes it sound like the alternative is a privacy violation, as opposed to helpful version checks and anonymous telemetry.

Comment (eliecharra) [reply build/build.go]: It can make sense yeah, what about `enableUsageReport` ?

Comment (craigfurman) [reply build/build.go]: Slight tweak: `enableUsageReporting`? I still like `enableTelemetry` 😁 

totally up to you though

**Pull request merged**
